zend/constant.cpp

```diff
@@ -0,0 +1,116 @@
+/**
+ *  constant.cpp
+ *
+ *  This file holds the implementation for the Php::constant(),
+ *  Php::define() and Php::defiend() functions
+ *
+ *  @author andot <https://github.com/andot>
+ */
+
+#include "includes.h"
+
+namespace Php {
+
+    /**
+     * Get the value of a constant
+     *  @param  constant_name
+     *  @return value
+     */
+    Php::Value constant(const std::string &constant_name) {
+        Php::Value result;
+        // we need the tsrm_ls variable
+        TSRMLS_FETCH();
+
+        if (!zend_get_constant_ex(constant_name.c_str(), constant_name.length(), result._val, NULL, ZEND_FETCH_CLASS_SILENT TSRMLS_CC)) {
+            php_error_docref(NULL TSRMLS_CC, E_WARNING, "Couldn't find constant %s", constant_name.c_str());
+            return nullptr;
+        }
+
+        return result;
+    }
+    /**
+     * Define a new constant
+     *  @param  constant_name
+     *  @param  value
+     *  @param  case_insensitive
+     *  @return bool
+     */
+    bool define(const std::string &constant_name, const Php::Value &value, bool case_insensitive) {
+        char *name = (char *)constant_name.c_str();
+        int name_len = constant_name.length();
+        zval *val = value._val;
+        zval *val_free = nullptr;
+        zend_constant c;
+        // we need the tsrm_ls variable
+        TSRMLS_FETCH();
+        /* class constant, check if there is name and make sure class is valid & exists */
+        if (zend_memnstr(name, (char *)"::", sizeof("::") - 1, name + name_len)) {
+            zend_error(E_WARNING, "Class constants cannot be defined or redefined");
+            return false;
+        }
+    repeat:
```

Comment:
It could be better with recursive call through overloaded functions to loop around HashIterator